File: android/app/src/main/java/com/volleycut/nativeanalysis/SuppressionModelRunner.java
package com.volleycut.nativeanalysis;

import android.content.Context;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/** Exact single-head runner for the frozen overlap-safe suppression specialist. */
final class SuppressionModelRunner {
    record Result(float[] probabilities, List<AnalysisTypes.Interval> decodedIntervals) {}
    private record Head(float[] mean, float[] scale, float[] weights, float bias) {}
    private record Decoder(
            double smoothingSeconds,
            double enterThreshold,
            double exitThreshold,
            double minLiveSeconds,
            double bridgeGapSeconds,
            double shortEventMinSeconds,
            double shortEventThreshold
    ) {}

    private final Head head;
    private final Decoder decoder;
    private final double analysisFps;

    SuppressionModelRunner(Context context) throws IOException, JSONException {
        this(readAsset(context));
    }

    SuppressionModelRunner(JSONObject json) throws JSONException {
        validateIdentity(json);
        validateFeatures(json.getJSONArray("featureNames"));
        analysisFps = json.getDouble("analysisFps");
        JSONObject values = json.getJSONObject("head");
        int dimensions = FeatureSchema.contextualNames().size();
        head = new Head(
                floats(values.getJSONArray("mean"), dimensions),
                floats(values.getJSONArray("scale"), dimensions),
                floats(values.getJSONArray("weights"), dimensions),
                (float) values.getDouble("bias")
        );
        JSONObject held = json.getJSONObject("decoder");
        decoder = new Decoder(
                held.getDouble("smoothing_seconds"),
                held.getDouble("enter_threshold"),
                held.getDouble("exit_threshold"),
                held.getDouble("min_live_seconds"),
                held.getDouble("bridge_gap_seconds"),
                held.getDouble("short_event_min_seconds"),
                held.getDouble("short_event_threshold")
        );
        validateHeldDecoder(decoder);
    }

    private static JSONObject readAsset(Context context) throws IOException, JSONException {
        try (InputStream input = context.getAssets().open(
                FeatureSchema.SUPPRESSION_MODEL_ID + ".json")) {
            ByteArrayOutputStream output = new ByteArrayOutputStream();
            byte[] buffer = new byte[8192];
            int count;
            while ((count = input.read(buffer)) >= 0) output.write(buffer, 0, count);
            return new JSONObject(output.toString(StandardCharsets.UTF_8));
        }
    }

    Result run(double[] times, float[] contextual, double durationSeconds) {
        int rows = times.length;
        int dimensions = head.weights.length;
        if (contextual.length != rows * dimensions) {
            throw new IllegalArgumentException("Suppression matrix shape mismatch");
        }
        float[] probabilities = new float[rows];
        for (int row = 0; row < rows; row++) {
            int offset = row * dimensions;
            float logit = head.bias;
            for (int column = 0; column < dimensions; column++) {
                logit += ((contextual[offset + column] - head.mean[column]) / head.scale[column])
                        * head.weights[column];
            }
            probabilities[row] = (float) (1 / (1 + Math.exp(-Math.max(-30, Math.min(30, logit)))));
        }
        return new Result(probabilities, decode(times, probabilities, durationSeconds));
    }

    private List<AnalysisTypes.Interval> decode(
            double[] times, float[] probabilities, double durationSeconds
    ) {
        double fps = effectiveFps(times, analysisFps);
        int smoothingSamples = Math.max(1, pythonRound(decoder.smoothingSeconds * fps));
        float[] smoothed = smooth(probabilities, smoothingSamples);
        boolean[] live = new boolean[smoothed.length];
        boolean active = false;
        for (int index = 0; index < smoothed.length; index++) {
            if (!active && smoothed[index] >= decoder.enterThreshold) active = true;
            else if (active && smoothed[index] < decoder.exitThreshold) active = false;
            live[index] = active;
        }
        int bridgeSamples = Math.max(0, pythonRound(decoder.bridgeGapSeconds * fps));
        for (int[] run : runs(live, false)) {
            if (run[0] > 0 && run[1] < live.length && run[1] - run[0] <= bridgeSamples) {
                for (int index = run[0]; index < run[1]; index++) live[index] = true;
            }
        }
        int minimumLive = Math.max(1, pythonRound(decoder.minLiveSeconds * fps));
        int shortMinimum = Math.max(1, pythonRound(decoder.shortEventMinSeconds * fps));
        for (int[] run : runs(live, true)) {
            float peak = -Float.MAX_VALUE;
            for (int index = run[0]; index < run[1]; index++) peak = Math.max(peak, smoothed[index]);
            if (run[1] - run[0] < minimumLive
                    && !(run[1] - run[0] >= shortMinimum && peak >= decoder.shortEventThreshold)) {
                for (int index = run[0]; index < run[1]; index++) live[index] = false;
            }
        }
        double sampleWidth = 1 / fps;
        ArrayList<AnalysisTypes.Interval> result = new ArrayList<>();
        for (int[] run : runs(live, true)) {
            double start = Math.max(0, times[run[0]] - sampleWidth / 2);
            double end = Math.min(durationSeconds, times[run[1] - 1] + sampleWidth / 2);
            if (end <= start) continue;
            float score = 0;
            for (int index = run[0]; index < run[1]; index++) score += smoothed[index];
            result.add(new AnalysisTypes.Interval(start, end, score / (run[1] - run[0])));
        }
        return List.copyOf(result);
    }

    private static void validateIdentity(JSONObject json) throws JSONException {
        if (json.getInt("schemaVersion") != 1
                || !FeatureSchema.SUPPRESSION_MODEL_ID.equals(json.getString("modelId"))
                || !FeatureSchema.SUPPRESSION_ARTIFACT_SHA256.equals(json.getString("artifactSha256"))
                || !FeatureSchema.SUPPRESSION_WEIGHTS_SHA256.equals(json.getString("weightsSha256"))
                || !FeatureSchema.SUPPRESSION_DECODER_VERSION.equals(json.getString("decoderVersion"))) {
            throw new JSONException("Suppression model identity does not match the product contract");
        }
    }

    private static void validateFeatures(JSONArray names) throws JSONException {
        List<String> expected = FeatureSchema.contextualNames();
        if (names.length() != expected.size()) throw new JSONException("Suppression feature count mismatch");
        for (int index = 0; index < names.length(); index++) {
            if (!expected.get(index).equals(names.getString(index))) {
                throw new JSONException("Suppression feature signature differs at column " + index);
            }
        }
    }

    private static void validateHeldDecoder(Decoder value) throws JSONException {
        if (value.smoothingSeconds != 1.0 || value.enterThreshold != .75
                || value.exitThreshold != .65 || value.minLiveSeconds != .5
                || value.bridgeGapSeconds != .5 || value.shortEventMinSeconds != .25
                || value.shortEventThreshold != .9) {
            throw new JSONException("Suppression decoder is not the held production decoder");
        }
    }

    private static float[] floats(JSONArray array, int expected) throws JSONException {
        if (array.length() != expected) throw new JSONException("Suppression vector length mismatch");
        float[] result = new float[expected];
        for (int index = 0; index < expected; index++) result[index] = (float) array.getDouble(index);
        return result;
    }

    private static float[] smooth(float[] values, int requestedWindow) {
        if (values.length == 0 || requestedWindow <= 1) return values.clone();
        int window = Math.min(requestedWindow, values.length);
        int leftPadding = window / 2;
        float[] result = new float[values.length];
        for (int output = 0; output < values.length; output++) {
            float sum = 0;
            for (int kernel = 0; kernel < window; kernel++) {
                int source = Math.max(0, Math.min(values.length - 1, output + kernel - leftPadding));
                sum += values[source] / window;
            }
            result[output] = sum;
        }
        return result;
    }

    private static List<int[]> runs(boolean[] values, boolean expected) {
        ArrayList<int[]> result = new ArrayList<>();
        Integer start = null;
        for (int index = 0; index < values.length; index++) {
            if (values[index] == expected && start == null) start = index;
            if (values[index] != expected && start != null) {
                result.add(new int[]{start, index});
                start = null;
            }
        }
        if (start != null) result.add(new int[]{start, values.length});
        return result;
    }

    private static int pythonRound(double value) {
        int floor = (int) Math.floor(value);
        double fraction = value - floor;
        if (fraction < .5) return floor;
        if (fraction > .5) return floor + 1;
        return floor % 2 == 0 ? floor : floor + 1;
    }

    private static double effectiveFps(double[] times, double fallback) {
        if (times.length <= 1) return fallback;
        double[] differences = new double[times.length - 1];
        for (int index = 1; index < times.length; index++) {
            differences[index - 1] = times[index] - times[index - 1];
        }
        java.util.Arrays.sort(differences);
        double median = differences[differences.length / 2];
        return Double.isFinite(median) && median > 0 ? 1 / median : fallback;
    }
}

File: ios/Sources/VolleyCore/SuppressionModelRunner.swift
import Foundation

/// Frozen overlap-safe specialist, including Android's upper-median cadence and division smoothing.
public struct SuppressionModelRunner: Sendable {
    public static let modelId = "suppression-overlap-exclusion-retrained"
    public static let artifactSha256 = "39eddf58163901930434ea422a802686ae921ae1e8fe23c20a3c12e5f453da93"
    public static let weightsSha256 = "a943749b69c98a1bc926f8efc9fe60c67c1226534fe09a892c632519217aa3bb"
    public static let decoderVersion = "held-production-suppression-decoder-v1"
    public static let assetSha256 = "02274d0f17b89cd54ea24da7d1665a6475e48dc0f554d06092e9ef892332d4f1"

    public struct Result: Codable, Sendable {
        public let probabilities: [Float]
        public let decodedIntervals: [Interval]
    }
    private struct Head: Decodable, Sendable { let mean: [Float], scale: [Float], weights: [Float], bias: Float }
    private struct Bundle: Decodable, Sendable {
        let schemaVersion: Int, modelId: String, artifactSha256: String, weightsSha256: String, decoderVersion: String
        let analysisFps: Double, featureNames: [String], head: Head, decoder: [String: Double]
    }
    private let bundle: Bundle
    public init(data: Data, expectedFeatureNames: [String] = FeatureSchema.contextualNames) throws {
        let value = try JSONDecoder().decode(Bundle.self, from: data)
        guard value.schemaVersion == 1, value.modelId == Self.modelId,
              value.artifactSha256 == Self.artifactSha256, value.weightsSha256 == Self.weightsSha256,
              value.decoderVersion == Self.decoderVersion, value.featureNames == expectedFeatureNames,
              !expectedFeatureNames.isEmpty, value.analysisFps.isFinite, value.analysisFps > 0 else {
            throw AnalysisError.invalid("Suppression identity or feature signature does not match the product contract")
        }
        let held: [String: Double] = ["smoothing_seconds": 1, "enter_threshold": 0.75, "exit_threshold": 0.65,
            "min_live_seconds": 0.5, "bridge_gap_seconds": 0.5, "short_event_min_seconds": 0.25, "short_event_threshold": 0.9]
        guard held.allSatisfy({ value.decoder[$0.key] == $0.value }) else {
            throw AnalysisError.invalid("Suppression decoder differs from the held production decoder")
        }
        let head = value.head
        guard head.mean.count == expectedFeatureNames.count, head.scale.count == expectedFeatureNames.count,
              head.weights.count == expectedFeatureNames.count, head.mean.allSatisfy(\.isFinite), head.weights.allSatisfy(\.isFinite),
              head.scale.allSatisfy({ $0.isFinite && $0 > 0 }), head.bias.isFinite else {
            throw AnalysisError.invalid("Invalid suppression model vectors")
        }
        bundle = value
    }
    public func run(times: [Double], contextual: [Float], duration: Double) throws -> Result {
        guard duration.isFinite, duration >= 0 else { throw AnalysisError.invalid("Invalid suppression duration") }
        if times.isEmpty {
            guard contextual.isEmpty else { throw AnalysisError.invalid("Suppression matrix shape mismatch") }
            return Result(probabilities: [], decodedIntervals: [])
        }
        try FeatureMath.validate(times: times, values: contextual, columns: bundle.featureNames.count)
        guard times.last! <= duration else { throw AnalysisError.invalid("Suppression sample exceeds duration") }
        let head = bundle.head, width = head.weights.count
        let probabilities: [Float] = times.indices.map { row in
            var logit = head.bias
            for column in 0..<width {
                let normalized = (contextual[row * width + column] - head.mean[column]) / head.scale[column]
                logit += normalized * head.weights[column]
            }
            return Float(1 / (1 + exp(-Double(max(-30, min(30, logit))))))
        }
        return Result(probabilities: probabilities, decodedIntervals: decode(times, probabilities, duration))
    }
    private func runs(_ values: [Bool], _ expected: Bool) -> [Range<Int>] {
        var result: [Range<Int>] = [], start: Int?
        for i in values.indices {
            if values[i] == expected && start == nil { start = i }
            if values[i] != expected, let lower = start { result.append(lower..<i); start = nil }
        }
        if let start { result.append(start..<values.count) }
        return result
    }
    private func decode(_ times: [Double], _ probabilities: [Float], _ duration: Double) -> [Interval] {
        var fps = bundle.analysisFps
        if times.count > 1 {
            let differences = zip(times, times.dropFirst()).map { $1 - $0 }.sorted()
            let median = differences[differences.count / 2]
            if median.isFinite && median > 0 { fps = 1 / median }
        }
        func samples(_ seconds: Double) -> Int {
            let count = (seconds * fps).rounded(.toNearestOrEven)
            return count >= Double(Int.max) ? Int.max : Int(count)
        }
        let window = min(probabilities.count, max(1, samples(1)))
        var smoothed = probabilities
        if window > 1 {
            for output in probabilities.indices {
                var sum: Float = 0
                for kernel in 0..<window {
                    let source = max(0, min(probabilities.count - 1, output + kernel - window / 2))
                    sum += probabilities[source] / Float(window)
                }
                smoothed[output] = sum
            }
        }
        var active = false
        var live = smoothed.map { probability in
            if !active && Double(probability) >= 0.75 { active = true }
            else if active && Double(probability) < 0.65 { active = false }
            return active
        }
        let bridge = max(0, samples(0.5))
        for run in runs(live, false) where run.lowerBound > 0 && run.upperBound < live.count && run.count <= bridge {
            for i in run { live[i] = true }
        }
        let minimum = max(1, samples(0.5)), shortMinimum = max(1, samples(0.25))
        for run in runs(live, true) {
            if run.count < minimum && !(run.count >= shortMinimum && Double(smoothed[run].max()!) >= 0.9) {
                for i in run { live[i] = false }
            }
        }
        return runs(live, true).compactMap { run in
            let start = max(0, times[run.lowerBound] - 0.5 / fps), end = min(duration, times[run.upperBound - 1] + 0.5 / fps)
            guard end > start else { return nil }
            var score: Float = 0; for i in run { score += smoothed[i] }
            return Interval(start: start, end: end, confidence: score / Float(run.count))
        }
    }
}
